feat: implement Get and List ModelLine cli commands, and help options for each subcommands

roaminggypsy opened the pull request.

wfa-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)
<!-- Reviewable:end -->

roaminggypsy commented:
This was not mentioned in the CLI spec doc, but Craig and I agreed that these two commands are needed.

SanjayVas reviewed (approved):
> Reviewed 2 of 3 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @stevenwarejones)



<!-- Sent from Reviewable.io -->

roaminggypsy commented:
The latest commit added two new changes

- added 'mixinStandardHelpOptions = true' to enable --help and -h for each individual commands
- changed --kingdom-public-api-target specification and check so that a missing target will not block 'help' for each subcommands

Before:
```
bazel run //src/main/kotlin/org/wfanet/measurement/kingdom/deploy/tools:ModelRepository 
-- model-lines create --help

Missing required option: '--kingdom-public-api-target=<target>'
```


After:
```
bazel run //src/main/kotlin/org/wfanet/measurement/kingdom/deploy/tools:ModelRepository 
-- model-lines create --help

Usage: model-repository model-lines create [-hV] [--active-end-time=<endTime>]
       --active-start-time=<startTime> [--description=<modelLineDescription>]
       [--display-name=<modelLineDisplayName>]
       [--holdback-model-line=<holdbackModelLineName>]
       --parent=<parentModelSuite> --population=<populationName>
       --type=<modelLineType>
Create a ModelLine, a ModelRelease, and a ModelRollout)
      --active-end-time=<endTime>
                  The end (exclusive) of the time range when the ModelLine is
                    active
      --active-start-time=<startTime>
                  The start of the time range when this ModelLine is active
      --description=<modelLineDescription>
                  Human-readable description of usage of the ModelLine
      --display-name=<modelLineDisplayName>
                  Human-readable nickname for the ModelLine
  -h, --help      Show this help message and exit.
      --holdback-model-line=<holdbackModelLineName>
                  API resource name of the holdback ModelLine that is used to
                    generate reports when this ModelLine presents outages. Only
                    PROD ModelLine can have this set.
      --parent=<parentModelSuite>
                  API resource name of the parent ModelSuite
      --population=<populationName>
                  API resource name of the Population referenced by ModelRelease
      --type=<modelLineType>
                  Type of the ModelLine
  -V, --version   Print version information and exit.
```

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OU7MVoyDlHHb2biyEfz:b-pzsmm3) (no related file):*
> added 'mixinStandardHelpOptions = true' to enable --help and -h for each individual commands

We want to use the help command instead. You do this by adding `HelpCommand::class` as a subcommand.

> changed --kingdom-public-api-target specification and check so that a missing target will not block 'help' for each subcommands

It's fine for it to block help on the subcommand.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier, @roaminggypsy, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OU7MyfKDGnukbrxoGV7:bcuvcsv) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

> added 'mixinStandardHelpOptions = true' to enable --help and -h for each individual commands

We want to use the help command instead. You do this by adding `HelpCommand::class` as a subcommand.

> changed --kingdom-public-api-target specification and check so that a missing target will not block 'help' for each subcommands

It's fine for it to block help on the subcommand.
</blockquote></details>

src/main/kotlin/org/wfanet/measurement/common/tools/OpenIdProvider.kt has an example of using `CommandLine.HelpCommand::class` along with printing usage on the top-level run method.


<!-- Sent from Reviewable.io -->

kungfucraig reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier, @SanjayVas, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OUArg0_88psa3Lfb4ry:bjz3b23) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

src/main/kotlin/org/wfanet/measurement/common/tools/OpenIdProvider.kt has an example of using `CommandLine.HelpCommand::class` along with printing usage on the top-level run method.
</blockquote></details>

If there is a way to not block help on the subcommand that's highly preferred. Consider that you should be able to ask for help for all things without even knowing what your credentials might be or whether you have access to any server at all.


<!-- Sent from Reviewable.io -->

roaminggypsy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier, @SanjayVas, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OUB0Vhq26s0N-Ya-SQs:butghxq) (no related file):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

If there is a way to not block help on the subcommand that's highly preferred. Consider that you should be able to ask for help for all things without even knowing what your credentials might be or whether you have access to any server at all.
</blockquote></details>

One way to do it is remove 'required = true' for --kingdom-public-api-target, but require it to be no null during channel creation. This is in the second bullet point in my second comment.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @kungfucraig, @Marco-Premier, @roaminggypsy, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OUBJnxm0xxQimOBypNk:b-2an517) (no related file):*
<details><summary><i>Previously, roaminggypsy (Lin) wrote…</i></summary><blockquote>

One way to do it is remove 'required = true' for --kingdom-public-api-target, but require it to be no null during channel creation. This is in the second bullet point in my second comment.
</blockquote></details>

Requiring all required parent options to be specified for any subcommand is consistent with how all of our other CLIs behave and is the expected behavior from picocli. If it's something we want to address, we should figure out a general approach and apply it consistently. That is to say, don't try to do it just for this one CLI in this one PR.


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @kungfucraig, @Marco-Premier, @roaminggypsy, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OUBLK9VDZLGfvZaHJBR:b7ww5i1) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Requiring all required parent options to be specified for any subcommand is consistent with how all of our other CLIs behave and is the expected behavior from picocli. If it's something we want to address, we should figure out a general approach and apply it consistently. That is to say, don't try to do it just for this one CLI in this one PR.
</blockquote></details>

Note that help commands don't check required options at the same level, just the parent level. For example, `help model-providers` doesn't need any required options set, but `model-providers help list` does. Unfortunately I don't spot a built-in way to get picocli to avoid `MissingParameterException` at all levels. Regardless, my point still stands that we shouldn't try to address that in this PR.


<!-- Sent from Reviewable.io -->

kungfucraig reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier, @SanjayVas, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OUB_bQB77WDRgCbFIWk:bppxtzv) (no related file):*
<details><summary><i>Previously, SanjayVas (Sanjay Vasandani) wrote…</i></summary><blockquote>

Note that help commands don't check required options at the same level, just the parent level. For example, `help model-providers` doesn't need any required options set, but `model-providers help list` does. Unfortunately I don't spot a built-in way to get picocli to avoid `MissingParameterException` at all levels. Regardless, my point still stands that we shouldn't try to address that in this PR.
</blockquote></details>

I agree that it's a separate PR. 

And if it's not easy to do, which it looks like it isn't (which is sad), then we should not do it now. Rather we should file an issue (which applies to all CLIs) and see if we can figure this out another time.


<!-- Sent from Reviewable.io -->

roaminggypsy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier, @SanjayVas, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OUC6g22B-fxy-Fn6Gws:b-3i1dts) (no related file):*
<details><summary><i>Previously, kungfucraig (Craig Wright) wrote…</i></summary><blockquote>

I agree that it's a separate PR. 

And if it's not easy to do, which it looks like it isn't (which is sad), then we should not do it now. Rather we should file an issue (which applies to all CLIs) and see if we can figure this out another time.
</blockquote></details>

I made the target flag required again, and add some notes about leaf command help to the cli documentation.


<!-- Sent from Reviewable.io -->

roaminggypsy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: 2 of 4 files reviewed, 1 unresolved discussion (waiting on @Marco-Premier, @SanjayVas, and @stevenwarejones)

___
*[a discussion](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535#-OU7MVoyDlHHb2biyEfy:-OUC89ma49erHQPrSaRg:b-2pmcmr) (no related file):*
<details><summary><i>Previously, roaminggypsy (Lin) wrote…</i></summary><blockquote>

I made the target flag required again, and add some notes about leaf command help to the cli documentation.
</blockquote></details>

Filed the issue #2544   


<!-- Sent from Reviewable.io -->

SanjayVas reviewed (approved):
> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/2535)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @Marco-Premier and @stevenwarejones)



<!-- Sent from Reviewable.io -->
